### DISCUSSION TITLE
Factoring in existing facilities

### DISCUSSION THREAD

**Discussion by rajesvariparasa** [pysal/spopt#218] [Ideas] [answered]
Hello,

Thanks for this library. I am testing it for a typical use case - coverage of demand locations by primary healthcare centres. I checked the examples and they seem to consider all facility locations as candidate sites. However, in my case, some of the facilities are already constructed and cannot be removed/relocated. I only need to add more facilities in the most optimal way. Can you tell me how I should tweak the examples discussed [here](https://pysal.org/spopt/notebooks/facloc-real-world.html#) to do that? Thanks!

**Comment by gegen07** [comment]
Hey @rajesvariparasa, great to hear this from you. One solution that I thought of is to change the problem a little bit. Firstly, you can see which demand points are not covered by your facilities already constructed. Then consider all those points to assign new facilities sites in the most optimal way. 

What do you think about this problem? @jGaboardi @qszhao

**Comment by jGaboardi** [comment]
This is an interesting problem. I'll move it to a Discussion.

**Comment by jGaboardi** [comment]
@rajesvariparasa

1. What's the original coverage model you implement? The LSCP or MCLP?
2. Are the client/demand locations added or subtracted?

One way to solve this would be to stipulate the decision variable representing the facilities is always set to 1 and we can do this by simply adding an extra constraints to the formulation. For example, in a set of facilities <img src="https://render.githubusercontent.com/render/math?math=\{y_0, y_1, y_2, y_3\}">, if facilities <img src="https://render.githubusercontent.com/render/math?math=\{y_0, y_1\}"> are already located then a constraint can be added to the formulation for each one as <img src="https://render.githubusercontent.com/render/math?math=y_0=1"> and <img src="https://render.githubusercontent.com/render/math?math=y_1=1">, which ensures that they are selected in the model.

We do not (yet?) have this functionality automated in `spopt`.

**Reply by rajesvariparasa** [reply]
Hi! 
1. I plan to implement LSCP first. 
2. Not sure I understand this question. Do you mean if the demand locations are allocated or deallocated to the facility site to reach optimality? So for now, no facility has reached its maximum capacity to serve to require any sort of decrease in the demand (here, population). However, there are demand sites that are not covered by the facilities because of the far distances. (I didn't realise it was possible to add this constraint of maximum demand besides the minimum distance). So I am guessing, the answer would be that the population numbers will be added up. Again, not sure if this is what you were asking?

And yes, I can pass another attribute column in the facilities dataset with a Boolean value for whether or not it should be included in the solution.
 
To give more context (not sure if it adds any more info), the demand sites are the slums in the city with a varying population. And the slum dwellers are the target beneficiaries of the urban primary healthcare centres in the city. While the facilities tend to be more concentrated towards the centre of the city, the slums come up towards the peripheries due to a bunch of reasons. We are trying to demonstrate that better spatial planning of healthcare facilities can help bring the people located at the peripheries under the formal healthcare network.

Now as per the official guidelines, every slum should ideally have a facility within 1 km of distance. But there are populations that are even as far as 9 km from the nearest facility. So we are trying to come up with potential locations for new facilities so that all slum locations are covered.


@gegen07 that does make sense -- I ll try this.

**Reply by qszhao** [reply]
Hi,

Thanks for your message and for trying our packages for your own research. I would say what you want to achieve is a very typical LSCP problem with predefined facilities. I would say try James's suggestions first and see what you can get. It may be tricky to use Germano's methods since that might result in locating more facilities than we need. 

This is a paper doing a very similar application to what you want to do: 
Murray, A. T. (2013). Optimising the spatial location of urban fire stations. Fire Safety Journal, 62, 64–71. https://doi.org/10.1016/j.firesaf.2013.03.002

@jGaboardi @gegen07 We will certainly need to think of a better way to do it. I believe in ArcGIS Network Analyst, there is a setting for users to define which stations are pre-existing. 

Best,
Qunshan

**Comment by jGaboardi** [comment, marked as answer]
@rajesvariparasa A solution for stipulating existing facilities has been implemented in #231 by @gegen07.
